NAS-100704 / 11.3 / Add method to retrieve plugin versions

sonicaj opened the pull request:
This commit adds the ability to retrieve plugin versions via packagesite defined for each plugin. As packagesite can differ for each plugin, the only reliable way to make sure that version is valid is to use packagesite.txz and extract version from there.

For now this isn't exposed to CLI because i don't see a need to the why here ( time to execute one of the main reasons ). If we get a feature request or in the future, we can easily add a flag to expose this in the CLI.

sonicaj committed:
Add method to retrieve plugin versions

This commit adds the ability to retrieve plugin versions via packagesite defined for each plugin. As packagesite can differ for each plugin, we retrieve each packagesite's ports data and isolate version from there.

william-gr reviewed (changes requested):
We need to improve this for speed.

Sequentially fetching each plugin json will be very slow, we should investigate either reuse the same connection to grab more files or perhaps fetch all at once using git (since we need the INDEX anayway, its probably better to just use git to get it all).

Same for packagesite, get a unique set and have a common method to fetch and extract it using a thread pool.

sonicaj committed:
Separete clone/pull git repo method

This commit separates clone/pull method to update git repo so it can be used independently.

skarekrow commented:
For completeness sake, adding vlad to this (my review is the same as williams)

sonicaj commented:
@william-gr there can be a case when the pool hasn't been activated and we want to retrieve plugin versions. I suppose in this case we can check for activation of pool and if we don't find one, we retrieve the plugins files in a manner which reduces speed.

skarekrow commented:
We shouldn't have a use case like that @sonicaj. iocage requires a pool to do any operations, having a few exceptions just creates confusion on where to store artifacts.

sonicaj commented:
@skarekrow the idea is that we do all of this as soon as middlewared boots and cache the results instead of doing this the first time pool is activated and user actually is at the plugins screen waiting for the plugins to load

skarekrow commented:
@sonicaj We can do something special in the mw for that. This is for all of iocage, which isn't a good fit for a compromise like that

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,137 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+        else:
+            git_server = self.server
+
+        git_working_dir = f"{self.iocroot}/.plugin_index"
```
Wont this require iocage to be activated to show a list of plugins?

sonicaj replied:
That is the whole idea as @skarekrow suggested. I have separated the functions, once we have the plugin index and plugin json files data, that data can be passed on to the new function which retrieves versions.

sonicaj committed:
Allow params for clone method

This commit adds a depth param to clone method which can be used to create shallow clones.

sonicaj committed:
Decouple cloning repo functionality

This commit decouples cloning repo functionality from plugin interface so it can be used in cases when even the pool hasn't been activated yet.

skarekrow commented on iocage_lib/ioc_common.py:
```diff
@@ -107,6 +108,78 @@ def logit(content, _callback=None, silent=False, exception=RuntimeError):
     _callback(content, exception)
 
 
+def clone_repo(ref, repo_url, destination, depth=None, callback=None):
```
I'm not so sure this belongs in common, after all, only the IOCPlugin module will ever be doing this.

sonicaj replied:
I can have this as a static method in the plugin interface, please advise what you think is better in this case. But it would seem weird calling `IOCPlugin._clone_repo`.

skarekrow replied:
Well perhaps a different method name is appropriate, but it really should live in that module. It’s a private method for just it

sonicaj replied:
Well this would be the exact invocation, `IOCPlugin._clone_repo(branch, repo_url, destination, depth, callback)` with params. Should i proceed ?

skarekrow replied:
It's fine to me, @william-gr do you have a better suggestion for a method name?

william-gr replied:
Looks ok to me

skarekrow reviewed (changes requested):
We should store a json manifest that has a list of the plugins we had in the INDEX at last invocation. If they match, we should skip the fetching of package versions and use what is in that json manifest to save a lot of time and bandwidth. Otherwise this would happen each time the user requests the INDEX.

sonicaj commented:
@skarekrow we will be caching the results in mw, so that i believe is not needed here. If we ever decide to expose this in CLI, i believe it needs to be fresh. Also maintaining a list of plugins is not enough, versions can still be updated ;)

skarekrow commented:
Fair enough :P

william-gr reviewed (commented):
I will wait to see how this will look in jail plugin before approve.

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+        ])
+
+        with concurrent.futures.ThreadPoolExecutor() as exc:
+            futures = exc.map(
```
futures seems like a bad name, they are not futures, but an iterable of results?

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+    def fetch_plugin_versions_from_plugin_index(plugins_index):
+        plugin_packagesite_mapping = IOCPlugin.fetch_plugin_packagesites([
+            v['packagesite'] for v in plugins_index.values()
+            if v.get('packagesite')
```
Is packagesite not required in plugin? If so what is the default we use and why its not in this list?

sonicaj replied:
Ideally it is required. This is just better safe then sorry approach

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                pass
+            else:
+                with tempfile.TemporaryDirectory() as td:
+                    tar = tarfile.open(packagesite_path)
```
doesnt tarfile have context manager?

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                packagesite = packagesite[:-1] if packagesite[-1] == '/' \
+                    else packagesite
+                plugin_pkgs = plugin_packagesite_mapping[packagesite]
+                for pkg in plugin_pkgs:
```
So we are iterating through every single package in each packagesite? Did I get this right?

That does not seem very good (O(n) for every plugin). We should come up with a way to make it O(log n) using a better structure.

This also likely means using two indexes because origin and package name, see the other discussion about origin.

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                                pass
+                            else:
+                                package_site_data[
+                                    line_data['origin']
```
So I am confused about using origin.

How does this handle flavors?

A single origin can have multiple flavors and renders multiple different package names.

What if I want to use "unison-nox11" package from net/unison origin?

Perhaps we should start thinking about requiring pkg name and not origin in plugins? Since thats all it can use, packages and not ports.

sonicaj replied:
You're right. This will right now "apparently" work because we have origin's listed in primary package and not flavours. Now different flavours might have different versions which give unreliable results. I'll revisit the plugin INDEX and make sure we use pkg names instead of origin.

william-gr commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                response = requests.get(f'{packagesite_url}/All/')
+                response.raise_for_status()
+
+                for pkg in re.findall(r'<a.*>(.*).txz</a>', response.text):
```
These are greedy, perhaps `r'<a.*?>(.*?).txz</a>'`, also wondering if you dont need re.S and re.M? i dont think is guaranteed to have them all on same line.

As even safer approach would be xml parsing (beautifulsoup).

sonicaj replied:
@william-gr i tried xml parsing first, regex is 2x times faster. The way we have it right now, the flags aren't needed, but better safe then sorry, i'll add those ;)

william-gr replied:
Just remembered  we probably dont want another dependency in iocage, so this should be fine :P

william-gr replied:
Perhaps `r'\S+.txz'` ?

sonicaj replied:
Yup, the regex flags are not needed in our case. Just a little more to the regex ;)

william-gr reviewed (dismissed):
Much better ;) thanks

william-gr reviewed (dismissed):
Well, remember the html comes from the http server, so the output may differ from server to server (apache, nginx, lighttpd).

sonicaj commented:
The content between the tags should remain same ;)

themylogin commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+        def download_parse_packagesite(packagesite_url):
+            package_site_data = {}
+            try:
+                response = requests.get(f'{packagesite_url}/All/')
```
TImeout?

themylogin commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+        def download_parse_packagesite(packagesite_url):
+            package_site_data = {}
+            try:
+                response = requests.get(f'{packagesite_url}/All/', timeout=120)
```
120?

sonicaj replied:
Should i decrease it to 60 perhaps ?

themylogin commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                response = requests.get(f'{packagesite_url}/All/', timeout=20)
+                response.raise_for_status()
+
+                for pkg in re.findall(r'<a.*>\s*(\S+).txz</a>', response.text):
```
Please use `html.parser` module instead

sonicaj replied:
I did, it's way slower

themylogin replied:
I've got it wrong. Let's use `etree.HTMLParser` from `lxml`

william-gr replied:
We cant add yet another dependency to iocage only for this

themylogin replied:
If lxml is not a dependency of iocage then fine, let's leave it like this @sonicaj

themylogin commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+                    revision_split = epoch_split[0].rsplit('_', 1)
+                    revision = \
+                        revision_split[1] if len(revision_split) == 2 else '0'
+                    package_site_data[pkg] = {
```
How about moving this (parsing `pkg` into dict) to to separate function and creating some unit tests?

themylogin commented on iocage_lib/ioc_plugin.py:
```diff
@@ -84,6 +84,125 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+
+        plugin_packagesite_mapping = {}
+        package_sites = set([
+            url[:-1] if url[-1] == '/' else url for url in package_sites
```
`rstrip("/")`?

sonicaj committed:
Add unit tests for parsing pkg name

This commit adds unit tests for parsing pkg name.

themylogin reviewed (dismissed):
Thank you!
